sway/tree/layout.c

```diff
@@ -166,9 +184,264 @@ void container_move_to(struct sway_container *container,
 	arrange_windows(new_parent, -1, -1);
 }
 
+static bool sway_dir_to_wlr(enum movement_direction dir,
+		enum wlr_direction *out) {
+	switch (dir) {
+	case MOVE_UP:
+		*out = WLR_DIRECTION_UP;
+		break;
+	case MOVE_DOWN:
+		*out = WLR_DIRECTION_DOWN;
+		break;
+	case MOVE_LEFT:
+		*out = WLR_DIRECTION_LEFT;
+		break;
+	case MOVE_RIGHT:
+		*out = WLR_DIRECTION_RIGHT;
+		break;
+	default:
+		return false;
+	}
+
+	return true;
+}
+
+static bool is_parallel(enum sway_container_layout layout,
+		enum movement_direction dir) {
+	switch (layout) {
+	case L_TABBED:
+	case L_STACKED:
+	case L_HORIZ:
+		return dir == MOVE_LEFT || dir == MOVE_RIGHT;
+	case L_VERT:
+		return dir == MOVE_UP || dir == MOVE_DOWN;
+	default:
+		return false;
+	}
+}
+
+static enum movement_direction invert_movement(enum movement_direction dir) {
+	switch (dir) {
+	case MOVE_LEFT:
+		return MOVE_RIGHT;
+	case MOVE_RIGHT:
+		return MOVE_LEFT;
+	case MOVE_UP:
+		return MOVE_DOWN;
+	case MOVE_DOWN:
+		return MOVE_LEFT;
+	default:
+		sway_assert(0, "This function expects left|right|up|down");
+		return MOVE_LEFT;
+	}
+}
+
+/* Gets the index of the most extreme member based on the movement offset */
+static int container_limit(struct sway_container *container, int offs) {
+	if (container->children->length == 0) {
+		return 0;
+	}
+	return offs < 0 ? 0 : container->children->length - 1;
```

Comment:
Inserting an item at index `container->children->length - 1` should insert before the last item?

Reply:
I mean, it should be possible to hit `container->children->length` without selecting the parent

Reply:
That's not what this function is for, read the comment.